### PR TITLE
First-draft infrastructure for evaluation

### PR DESCRIPTION
**Pull request by thanasibakis**
This PR's goal is to take a first stab at how our package will provide model evaluation functionality.

### `eval` subpackage

The package now has an `eval` subpackage. The idea here is that this will contain modules for the various classes of metrics we will want (e.g. on proportions directly, on samples, on lineage domination times). I have a simple MAE on proportions in the `eval.proportions` submodule to start us out.

### `demo`

The `exploration/demo` workspace has two model-fitting scripts now. The idea here is that the outputs of each are collected into some output folder, and pitted against each other. `exploration/demo/evaluate.py` does this, but hardcodes CSV paths for now.

## Other changes

- The `utils` submodule continues to grow with helper functions that I think are nice to have all in one place. Happy to hear other ideas here.
- I will slowly move away from the R dependency for visualizing things. The `visualize` submodule has a new function for plotting posterior proportion trajectories over time, replacing the R script I had before.
- I pushed a very small change to the simulation study script to resolve a todo, which could have waited for another PR but snuck into this one accidentally.
- I also somehow managed earlier to commit some scratch CRPS code to main at some point, so that's why you'll see `linmod/eval.py` deleted here.

### CODE REVIEW

File: exploration/demo/evaluate.py
Comment (swo) [inline exploration/demo/evaluate.py:10]: `Path()` is redundant here. I think the "elegant" solution would be to have the root be known, so that you could say
```python
pl.scan_csv(root_path / "data" / "metadata.csv")
```

File: exploration/demo/evaluate.py
Comment (swo) [inline exploration/demo/evaluate.py:11]: I think the better think here is `try_parse_dates=True` in the `scan_csv()`

File: exploration/demo/evaluate.py
Comment (swo) [inline exploration/demo/evaluate.py:13]: let's put this 90 as a global variable up top
```python
EVAL_PERIOD_DAYS = 90
```
or similar

will make it easier to refactor later

File: linmod/eval/proportions.py
Comment (swo) [inline linmod/eval/proportions.py:1]: I think, just put this as a module, rather than as a sub-package? Maybe one day we'd have so many different kinds of evaluations, but for now, I think this will just confuse me

Comment (swo) [reply linmod/eval/proportions.py:1]: like, mash this material into `linmod/eval.py`

File: linmod/utils.py
Comment (swo) [inline linmod/utils.py:30]: I think you could just write this as the `if over is None` bit, because you could always do
```python
my_expr.pipe(pl_softmax).over(pl.col('group'))
```
right?

File: exploration/simulation-study/main.py
Comment (swo) [inline exploration/simulation-study/main.py:18]: someday you should learn argparse ;)

**Review by swo — CHANGES_REQUESTED** [review]
I had a quick look-over. I think I'll need to dig in more.

Can we get some update to the readme to explain what order these things should be run in, what data are needed, etc.?

File: exploration/demo/evaluate.py
Comment (thanasibakis) [reply exploration/demo/evaluate.py:22]: I actually intend to move that entire filter statement into our global preprocessing script, since it happens every time I use the data anyway. Should I still make that change in this PR?

File: exploration/demo/evaluate.py
Comment (thanasibakis) [reply exploration/demo/evaluate.py:20]: It would be cleaner, but there is actually a functionality difference that makes me still consider the current method...

We want to ensure the date column sets non-compliant entries (for example, dates that are only month+year) to `null`. This is important, because I use the null filter to remove rows that don't have dates resolved to the day. `cast(..., strict=False)` does this.

`scan_csv(..., try_parse_dates=True)` fails in this scenario without also setting `ignore_errors=True`, which scares me

Comment (thanasibakis) [reply exploration/demo/evaluate.py:20]: I think I should document this somewhere, even as a comment. But I will probably do that when I move this date filtering into the global preprocessing script. At that point, a user loading the output of that script could totally use `try_parse_dates=True` without fail, I think

File: linmod/utils.py
Comment (thanasibakis) [reply linmod/utils.py:30]: Absolutely, and I realized this after the fact and have been wanting to change it, but decided to leave it to a separate PR to not clutter this one. But now that you point it out, I am very happy change it :)

File: exploration/simulation-study/main.py
Comment (thanasibakis) [reply exploration/simulation-study/main.py:18]: Added to wishlist :)

File: exploration/demo/evaluate.py
Comment (swo) [reply exploration/demo/evaluate.py:20]: interesting. agree, this could be documented (or even written as a separate function)

File: exploration/demo/evaluate.py
Comment (swo) [reply exploration/demo/evaluate.py:22]: my idea is more that "magic" numbers should be easy to find & identify. so it's less about where this bit is in the codebase, but more about how even the preprocessing scripts should have their parameters as easy to find & modify as possible (b/c I'm sure at some point we'll ask how the model performance would differ if we used a different evaluation period)

Comment (thanasibakis) [reply exploration/demo/evaluate.py:22]: Totally. I made the change in another (draft) PR that deals with more data-processing-oriented tasks, it looks like this:

https://github.com/CDCgov/cfa-viral-lineage-model/blob/8dd8b0ccbcec850e768a63b724fcb30d0263e865/linmod/data.py#L33

Does that look alright?

File: exploration/demo/evaluate.py
Comment (thanasibakis) [reply exploration/demo/evaluate.py:20]: https://github.com/CDCgov/cfa-viral-lineage-model/blob/8dd8b0ccbcec850e768a63b724fcb30d0263e865/linmod/data.py#L160-L163

File: exploration/demo/evaluate.py
Comment (swo) [reply exploration/demo/evaluate.py:22]: nice, good enough for now

**Comment by afmagee42** [comment]
I'm encountering a few issues that are outside the scope of this PR, but which are blocking me from checking that the code runs in my newly-cloned repo. I suggest that this PR is merged as soon as @swo approves, and I will make a new branch to see about cleaning those up.

**Comment by swo** [comment]
I'm blocked right away:

```
linmod-py3.10➜  demo git:(first-metric) ../../data/load_metadata.py > metadata.csv
Will load from https://data.nextstrain.org/files/ncov/open/metadata.tsv.zst

Downloading...Traceback (most recent call last):
  File "/home/ulp7/proj/cfa-viral-lineage-model/exploration/demo/../../data/load_metadata.py", line 108, in <module>
    data = load_metadata(
  File "/home/ulp7/proj/cfa-viral-lineage-model/exploration/demo/../../data/load_metadata.py", line 58, in load_metadata
    with urlopen(url) as response, save_path.open("wb") as out_file:
  File "/usr/lib/python3.10/pathlib.py", line 1119, in open
    return self._accessor.open(self, mode, buffering, encoding, errors,
FileNotFoundError: [Errno 2] No such file or directory: '.cache/data.nextstrain.org/files/ncov/open/metadata.tsv'
```

**Comment by afmagee42** [comment]
@swo see #12

**Review by swo — CHANGES_REQUESTED** [review]
It seems like a lot of things are in flux in other PRs. The important thing here seems to be the score interface. I'd like to #13 or something like it integrated here first.

**Review by swo — APPROVED** [review]
Many more things to do, but those are in other PRs now

**Pull request merged**
